decision/engine.go

```diff
@@ -349,49 +376,85 @@ func (e *Engine) taskWorkerExit() {
 // context is cancelled before the next Envelope can be created.
 func (e *Engine) nextEnvelope(ctx context.Context) (*Envelope, error) {
 	for {
-		nextTask := e.peerRequestQueue.PopBlock()
-		for nextTask == nil {
+		// Pop some tasks off the request queue
+		p, nextTasks, pendingBytes := e.peerRequestQueue.PopTasks(targetMessageSize)
+		for len(nextTasks) == 0 {
 			select {
 			case <-ctx.Done():
 				return nil, ctx.Err()
 			case <-e.workSignal:
-				nextTask = e.peerRequestQueue.PopBlock()
+				p, nextTasks, pendingBytes = e.peerRequestQueue.PopTasks(targetMessageSize)
 			case <-e.ticker.C:
+				// When a task is cancelled, the queue may be "frozen" for a
+				// period of time. We periodically "thaw" the queue to make
+				// sure it doesn't get stuck in a frozen state.
 				e.peerRequestQueue.ThawRound()
-				nextTask = e.peerRequestQueue.PopBlock()
+				p, nextTasks, pendingBytes = e.peerRequestQueue.PopTasks(targetMessageSize)
 			}
 		}
 
-		// with a task in hand, we're ready to prepare the envelope...
+		// While there are more tasks to process
+		msg := bsmsg.New(true)
+
+		// log.Debugf("  %s got %d tasks", lu.P(e.self), len(nextTasks))
+
+		// Amount of data in the request queue still waiting to be popped
+		msg.SetPendingBytes(int32(pendingBytes))
+
+		// Add DONT_HAVEs to the message
+		for _, c := range filterDontHaves(nextTasks) {
+			// log.Debugf("  make evlp %s->%s DONT_HAVE %s", lu.P(e.self), lu.P(p), lu.C(c))
+			msg.AddDontHave(c)
+		}
+
+		// Add HAVEs to the message
+		for _, c := range filterWantHaves(nextTasks) {
+			// log.Debugf("  make evlp %s->%s HAVE %s", lu.P(e.self), lu.P(p), lu.C(c))
+			msg.AddHave(c)
+		}
+
+		// Get requested blocks from the blockstore
+		blockTasks := filterWantBlocks(nextTasks)
 		blockCids := cid.NewSet()
-		for _, t := range nextTask.Tasks {
-			blockCids.Add(t.Identifier.(cid.Cid))
+		for _, t := range blockTasks {
+			blockCids.Add(t.Topic.(cid.Cid))
 		}
 		blks := e.bsm.getBlocks(ctx, blockCids.Keys())
 
-		msg := bsmsg.New(true)
-		for _, b := range blks {
-			msg.AddBlock(b)
+		for _, t := range blockTasks {
+			c := t.Topic.(cid.Cid)
+			blk := blks[c]
+			// If the block was not found (it has been removed)
+			if blk == nil {
+				// If the client requested DONT_HAVE, add DONT_HAVE to the message
+				if t.Data.(*taskData).SendDontHave {
```

Comment:
Maybe we should just send this anyways. We could have told the peer we _have_ the block and this is an unlikely edge-case (so it's not like we're going to be flooding peers or anything).

Reply:
Nevermind, we send "send don't have" anyways.